From a1cc0d605d2c190c678be8460d2b7be67d9e8f5a Mon Sep 17 00:00:00 2001
From: "Ai, Ting" <ting.a.ai@intel.com>
Date: Mon, 17 Jan 2022 15:48:30 +0800
Subject: [PATCH] Update policy for media provider

Defined p9fs2 in public folder
Allow mediaprovider to access 9p sharefolder

Tracked-On: OAM-100644
Signed-off-by: Ai, Ting <ting.a.ai@intel.com>
---
 prebuilts/api/30.0/private/app_neverallows.te          | 3 ++-
 prebuilts/api/30.0/private/compat/29.0/29.0.ignore.cil | 3 ++-
 prebuilts/api/30.0/private/genfs_contexts              | 1 +
 prebuilts/api/30.0/private/mediaprovider_app.te        | 4 ++++
 prebuilts/api/30.0/public/app.te                       | 3 +++
 prebuilts/api/30.0/public/file.te                      | 2 ++
 private/app_neverallows.te                             | 3 ++-
 private/compat/29.0/29.0.ignore.cil                    | 3 ++-
 private/genfs_contexts                                 | 1 +
 private/mediaprovider_app.te                           | 4 ++++
 public/app.te                                          | 3 +++
 public/file.te                                         | 2 ++
 12 files changed, 28 insertions(+), 4 deletions(-)

diff --git a/prebuilts/api/30.0/private/app_neverallows.te b/prebuilts/api/30.0/private/app_neverallows.te
index 115718700..7670e3c68 100644
--- a/prebuilts/api/30.0/private/app_neverallows.te
+++ b/prebuilts/api/30.0/private/app_neverallows.te
@@ -130,9 +130,10 @@ neverallow { all_untrusted_apps -mediaprovider } { cache_file cache_recovery_fil
 # World accessible data locations allow application to fill the device
 # with unaccounted for data. This data will not get removed during
 # application un-installation.
-neverallow { all_untrusted_apps -mediaprovider } {
+neverallow { all_untrusted_apps -mediaprovider -mediaprovider_app -untrusted_app_29} {
   fs_type
   -sdcard_type
+  -p9fs2
   file_type
   -app_data_file            # The apps sandbox itself
   -privapp_data_file
diff --git a/prebuilts/api/30.0/private/compat/29.0/29.0.ignore.cil b/prebuilts/api/30.0/private/compat/29.0/29.0.ignore.cil
index e4aaef5f4..87da98e9c 100644
--- a/prebuilts/api/30.0/private/compat/29.0/29.0.ignore.cil
+++ b/prebuilts/api/30.0/private/compat/29.0/29.0.ignore.cil
@@ -129,4 +129,5 @@
     vendor_service_contexts_file
     vendor_socket_hook_prop
     vendor_socket_hook_prop
-    virtual_ab_prop))
+    virtual_ab_prop
+    p9fs2))
diff --git a/prebuilts/api/30.0/private/genfs_contexts b/prebuilts/api/30.0/private/genfs_contexts
index d05e907c4..1c86ef435 100644
--- a/prebuilts/api/30.0/private/genfs_contexts
+++ b/prebuilts/api/30.0/private/genfs_contexts
@@ -323,3 +323,4 @@ genfscon functionfs / u:object_r:functionfs:s0
 genfscon usbfs / u:object_r:usbfs:s0
 genfscon binfmt_misc / u:object_r:binfmt_miscfs:s0
 genfscon bpf / u:object_r:fs_bpf:s0
+genfscon 9p / u:object_r:p9fs2:s0
diff --git a/prebuilts/api/30.0/private/mediaprovider_app.te b/prebuilts/api/30.0/private/mediaprovider_app.te
index 5881255c9..de7b0e7a7 100644
--- a/prebuilts/api/30.0/private/mediaprovider_app.te
+++ b/prebuilts/api/30.0/private/mediaprovider_app.te
@@ -47,3 +47,7 @@ allow mediaprovider_app proc_filesystems:file r_file_perms;
 
 #Allow MediaProvider to see if sdcardfs is in use
 get_prop(mediaprovider_app, storage_config_prop)
+
+# Talk to 9P sharefolder
+allow mediaprovider_app p9fs2:file create_file_perms;
+allow mediaprovider_app p9fs2:dir create_dir_perms;
diff --git a/prebuilts/api/30.0/public/app.te b/prebuilts/api/30.0/public/app.te
index e5b9fd670..3ecd7ec53 100644
--- a/prebuilts/api/30.0/public/app.te
+++ b/prebuilts/api/30.0/public/app.te
@@ -359,6 +359,9 @@ allow appdomain audioserver_tmpfs:file { getattr map read write };
 allow appdomain system_server_tmpfs:file { getattr map read write };
 allow appdomain zygote_tmpfs:file { map read };
 
+allow appdomain p9fs2:dir create_dir_perms;
+allow appdomain p9fs2:file create_file_perms;
+
 ###
 ### Neverallow rules
 ###
diff --git a/prebuilts/api/30.0/public/file.te b/prebuilts/api/30.0/public/file.te
index 7ed8baab6..a6f786f04 100644
--- a/prebuilts/api/30.0/public/file.te
+++ b/prebuilts/api/30.0/public/file.te
@@ -549,3 +549,5 @@ type audiohal_data_file, file_type, data_file_type, core_data_file_type;
 # Should be:
 #   type apk_data_file, file_type, data_file_type;
 neverallow fs_type file_type:filesystem associate;
+
+type p9fs2, fs_type, mlstrustedobject;
diff --git a/private/app_neverallows.te b/private/app_neverallows.te
index 115718700..7670e3c68 100644
--- a/private/app_neverallows.te
+++ b/private/app_neverallows.te
@@ -130,9 +130,10 @@ neverallow { all_untrusted_apps -mediaprovider } { cache_file cache_recovery_fil
 # World accessible data locations allow application to fill the device
 # with unaccounted for data. This data will not get removed during
 # application un-installation.
-neverallow { all_untrusted_apps -mediaprovider } {
+neverallow { all_untrusted_apps -mediaprovider -mediaprovider_app -untrusted_app_29} {
   fs_type
   -sdcard_type
+  -p9fs2
   file_type
   -app_data_file            # The apps sandbox itself
   -privapp_data_file
diff --git a/private/compat/29.0/29.0.ignore.cil b/private/compat/29.0/29.0.ignore.cil
index e4aaef5f4..87da98e9c 100644
--- a/private/compat/29.0/29.0.ignore.cil
+++ b/private/compat/29.0/29.0.ignore.cil
@@ -129,4 +129,5 @@
     vendor_service_contexts_file
     vendor_socket_hook_prop
     vendor_socket_hook_prop
-    virtual_ab_prop))
+    virtual_ab_prop
+    p9fs2))
diff --git a/private/genfs_contexts b/private/genfs_contexts
index d05e907c4..1c86ef435 100644
--- a/private/genfs_contexts
+++ b/private/genfs_contexts
@@ -323,3 +323,4 @@ genfscon functionfs / u:object_r:functionfs:s0
 genfscon usbfs / u:object_r:usbfs:s0
 genfscon binfmt_misc / u:object_r:binfmt_miscfs:s0
 genfscon bpf / u:object_r:fs_bpf:s0
+genfscon 9p / u:object_r:p9fs2:s0
diff --git a/private/mediaprovider_app.te b/private/mediaprovider_app.te
index 5881255c9..de7b0e7a7 100644
--- a/private/mediaprovider_app.te
+++ b/private/mediaprovider_app.te
@@ -47,3 +47,7 @@ allow mediaprovider_app proc_filesystems:file r_file_perms;
 
 #Allow MediaProvider to see if sdcardfs is in use
 get_prop(mediaprovider_app, storage_config_prop)
+
+# Talk to 9P sharefolder
+allow mediaprovider_app p9fs2:file create_file_perms;
+allow mediaprovider_app p9fs2:dir create_dir_perms;
diff --git a/public/app.te b/public/app.te
index e5b9fd670..3ecd7ec53 100644
--- a/public/app.te
+++ b/public/app.te
@@ -359,6 +359,9 @@ allow appdomain audioserver_tmpfs:file { getattr map read write };
 allow appdomain system_server_tmpfs:file { getattr map read write };
 allow appdomain zygote_tmpfs:file { map read };
 
+allow appdomain p9fs2:dir create_dir_perms;
+allow appdomain p9fs2:file create_file_perms;
+
 ###
 ### Neverallow rules
 ###
diff --git a/public/file.te b/public/file.te
index 7ed8baab6..a6f786f04 100644
--- a/public/file.te
+++ b/public/file.te
@@ -549,3 +549,5 @@ type audiohal_data_file, file_type, data_file_type, core_data_file_type;
 # Should be:
 #   type apk_data_file, file_type, data_file_type;
 neverallow fs_type file_type:filesystem associate;
+
+type p9fs2, fs_type, mlstrustedobject;
-- 
2.25.1

